File: mcp-server/handlers/calendar.py
#!/usr/bin/env python3
"""
MemOS MCP Server Calendar Handler

Handles memos_calendar tool for student mode.
Provides semester/week/day views of learning notes.
"""

from datetime import datetime, timedelta
from typing import Any

import httpx

from api_client import api_call_with_retry
from config import MEMOS_URL, MEMOS_USER, logger
from cube_manager import ensure_cube_registered
from mcp.types import TextContent

from handlers.utils import (
    api_error_response,
    cube_registration_error,
    get_cube_id_from_args,
)


def get_semester_dates(semester: str) -> tuple[datetime, datetime]:
    """Get start and end dates for a semester."""
    now = datetime.now()

    if semester == "current":
        year = now.year
        if now.month >= 9:  # Fall: Sep-Dec
            return datetime(year, 9, 1), datetime(year, 12, 31)
        elif now.month >= 2:  # Spring: Feb-Jun
            return datetime(year, 2, 1), datetime(year, 6, 30)
        else:  # Winter break
            return datetime(year - 1, 9, 1), datetime(year - 1, 12, 31)

    try:
        year_str, season = semester.split("-")
        year = int(year_str)
        if season.lower() == "spring":
            return datetime(year, 2, 1), datetime(year, 6, 30)
        elif season.lower() == "fall":
            return datetime(year, 9, 1), datetime(year, 12, 31)
        elif season.lower() == "summer":
            return datetime(year, 7, 1), datetime(year, 8, 31)
    except (ValueError, AttributeError):
        pass

    return datetime(now.year, now.month, 1), now


def get_week_dates(semester_start: datetime, week: int) -> tuple[datetime, datetime]:
    """Get start and end dates for a specific week."""
    week_start = semester_start + timedelta(weeks=week - 1)
    week_start = week_start - timedelta(days=week_start.weekday())
    week_end = week_start + timedelta(days=6)
    return week_start, week_end


def format_calendar_output(
    memories: list[dict],
    semester: str,
    course: str | None,
    week: int | None,
    view: str,
) -> str:
    """Format memories as calendar view."""
    lines = []

    # Header
    if course:
        lines.append(f"## 📅 {semester} - {course}")
    else:
        lines.append(f"## 📅 {semester} 学期笔记")
    lines.append("")

    if not memories:
        lines.append("_暂无笔记_")
        return "\n".join(lines)

    # Group by date
    by_date: dict[str, list[dict]] = {}
    for mem in memories:
        created = mem.get("created_at", mem.get("timestamp", ""))
        if isinstance(created, str) and created:
            date_str = created[:10]
        else:
            date_str = datetime.now().strftime("%Y-%m-%d")

        if date_str not in by_date:
            by_date[date_str] = []
        by_date[date_str].append(mem)

    sorted_dates = sorted(by_date.keys(), reverse=True)

    if view == "list":
        lines.append("| 日期 | 类型 | 内容摘要 |")
        lines.append("|------|------|---------|")

        for date_str in sorted_dates[:20]:
            for mem in by_date[date_str]:
                mem_type = mem.get("memory_type", mem.get("type", "NOTE"))
                content = mem.get("content", mem.get("key", ""))[:50]
                if len(content) == 50:
                    content += "..."
                content = content.replace("|", "\\|").replace("\n", " ")
                lines.append(f"| {date_str} | {mem_type} | {content} |")

    elif view == "week":
        if week:
            lines.append(f"### 第 {week} 周")
        lines.append("")

        weekdays = ["周一", "周二", "周三", "周四", "周五", "周六", "周日"]
        for date_str in sorted_dates[:7]:
            try:
                dt = datetime.strptime(date_str, "%Y-%m-%d")
                weekday = weekdays[dt.weekday()]
                lines.append(f"**{date_str} {weekday}**")
            except ValueError:
                lines.append(f"**{date_str}**")

            for mem in by_date[date_str]:
                mem_type = mem.get("memory_type", mem.get("type", "NOTE"))
                content = mem.get("content", mem.get("key", ""))[:60]
                content = content.replace("\n", " ")
                lines.append(f"- [{mem_type}] {content}")
            lines.append("")

    else:  # month view
        lines.append("| 日期 | 笔记数 | 类型分布 |")
        lines.append("|------|--------|---------|")

        for date_str in sorted_dates[:30]:
            mems = by_date[date_str]
            types = {}
            for m in mems:
                t = m.get("memory_type", "OTHER")
                types[t] = types.get(t, 0) + 1
            type_str = ", ".join(f"{k}:{v}" for k, v in types.items())
            lines.append(f"| {date_str} | {len(mems)} | {type_str} |")

    # Stats
    total = len(memories)
    type_counts = {}
    for mem in memories:
        t = mem.get("memory_type", "OTHER")
        type_counts[t] = type_counts.get(t, 0) + 1

    lines.append("")
    lines.append(f"📊 总计: {total} 条笔记")
    if type_counts:
        lines.append(f"   类型: {', '.join(f'{k}:{v}' for k, v in sorted(type_counts.items(), key=lambda x: -x[1]))}")

    return "\n".join(lines)


async def handle_memos_calendar(
    client: httpx.AsyncClient,
    arguments: dict[str, Any]
) -> list[TextContent]:
    """Handle memos_calendar tool call."""
    cube_id = get_cube_id_from_args(arguments)
    semester = arguments.get("semester", "current")
    course = arguments.get("course")
    week = arguments.get("week")
    view = arguments.get("view", "list")

    reg_success, reg_error = await ensure_cube_registered(client, cube_id)
    if not reg_success:
        return cube_registration_error(cube_id, reg_error)

    start_date, end_date = get_semester_dates(semester)
    if week:
        start_date, end_date = get_week_dates(start_date, int(week))

    params = {
        "user_id": MEMOS_USER,
        "mem_cube_id": cube_id,
        "limit": 100,
    }

    success, data, status = await api_call_with_retry(
        client, "GET", f"{MEMOS_URL}/memories", cube_id,
        params=params
    )

    if success and data:
        result_data = data.get("data", {})

        memories = []
        text_mems = result_data.get("text_mem", [])
        for cube_data in text_mems:
            memories_data = cube_data.get("memories", [])
            if isinstance(memories_data, dict) and "nodes" in memories_data:
                memories.extend(memories_data["nodes"])
            elif isinstance(memories_data, list):
                memories.extend(memories_data)

        # Filter by course
        if course:
            memories = [m for m in memories
                       if course.lower() in m.get("content", "").lower()
                       or course.lower() in str(m.get("tags", [])).lower()]

        # Filter by date range
        filtered = []
        for mem in memories:
            created = mem.get("created_at", "")
            if isinstance(created, str) and created:
                try:
                    dt = datetime.fromisoformat(created.replace("Z", "+00:00"))
                    if start_date <= dt.replace(tzinfo=None) <= end_date:
                        filtered.append(mem)
                except (ValueError, TypeError):
                    filtered.append(mem)
            else:
                filtered.append(mem)

        output = format_calendar_output(filtered, semester, course, week, view)
        return [TextContent(type="text", text=output)]

    elif data:
        return api_error_response("Calendar", data.get("message", "Unknown error"))
    else:
        return api_error_response("Calendar", f"HTTP {status}")

File: mcp-server-node/src/handlers/calendar.ts
/**
 * Calendar Handler
 *
 * memos_calendar - project mode and student mode
 */

import { MEMOS_URL, MEMOS_USER } from "../config.js";
import { apiCallWithRetry } from "../api-client.js";
import { ensureCubeRegistered } from "../cube-manager.js";
import type { TextContent, MemoryNode, SearchData } from "../types.js";
import { apiErrorResponse, cubeRegistrationError, getCubeIdFromArgs } from "./utils.js";

// ============================================================================
// Date Utilities
// ============================================================================

function getSemesterDates(semester: string): [Date, Date] {
  const now = new Date();

  if (semester === "current") {
    const year = now.getFullYear();
    const month = now.getMonth() + 1; // 1-based
    if (month >= 9) {
      return [new Date(year, 8, 1), new Date(year, 11, 31)];
    } else if (month >= 2) {
      return [new Date(year, 1, 1), new Date(year, 5, 30)];
    } else {
      return [new Date(year - 1, 8, 1), new Date(year - 1, 11, 31)];
    }
  }

  const parts = semester.split("-");
  if (parts.length === 2) {
    const year = parseInt(parts[0]);
    const season = parts[1].toLowerCase();
    if (!isNaN(year)) {
      if (season === "spring") return [new Date(year, 1, 1), new Date(year, 5, 30)];
      if (season === "fall") return [new Date(year, 8, 1), new Date(year, 11, 31)];
      if (season === "summer") return [new Date(year, 6, 1), new Date(year, 7, 31)];
    }
  }

  return [new Date(now.getFullYear(), now.getMonth(), 1), now];
}

function getWeekDates(semesterStart: Date, week: number): [Date, Date] {
  const weekStart = new Date(semesterStart.getTime() + (week - 1) * 7 * 24 * 60 * 60 * 1000);
  const dayOfWeek = weekStart.getDay();
  const monday = new Date(weekStart.getTime() - ((dayOfWeek + 6) % 7) * 24 * 60 * 60 * 1000);
  const sunday = new Date(monday.getTime() + 6 * 24 * 60 * 60 * 1000);
  return [monday, sunday];
}

// ============================================================================
// Project Timeline Formatter
// ============================================================================

function formatProjectTimeline(memories: MemoryNode[], cubeId: string): string {
  const lines: string[] = [];
  lines.push(`## 🗓️ Project Timeline — ${cubeId}`, "");

  const projectTypes = new Set(["MILESTONE", "DECISION", "FEATURE", "BUGFIX", "ERROR_PATTERN", "GOTCHA"]);

  const projectMems = memories.filter((m) => {
    const t = String((m as Record<string, unknown>).memory_type ?? (m as Record<string, unknown>).type ?? "").toUpperCase();
    return projectTypes.has(t);
  });

  if (projectMems.length === 0) {
    lines.push("_No project milestones or decisions found._", "");
    lines.push('Save memories with: memos_save(..., memory_type="MILESTONE")');
    return lines.join("\n");
  }

  // Group by year-month
  const byMonth: Record<string, typeof projectMems> = {};
  for (const mem of projectMems) {
    const created = String((mem as Record<string, unknown>).created_at ?? (mem as Record<string, unknown>).timestamp ?? "");
    const monthStr = created.length >= 7 ? created.slice(0, 7) : "unknown";
    if (!byMonth[monthStr]) byMonth[monthStr] = [];
    byMonth[monthStr].push(mem);
  }

  const sortedMonths = Object.keys(byMonth).sort().reverse();

  for (const monthStr of sortedMonths) {
    const mems = byMonth[monthStr];
    lines.push(`**${monthStr}**`);
    for (let i = 0; i < mems.length; i++) {
      const mem = mems[i];
      const memType = String((mem as Record<string, unknown>).memory_type ?? (mem as Record<string, unknown>).type ?? "NOTE").toUpperCase();
      const content = String((mem as Record<string, unknown>).content ?? mem.key ?? "");
      const summary = content.split("\n")[0].slice(0, 80);
      const connector = i === mems.length - 1 ? "└──" : "├──";
      lines.push(`  ${connector} [${memType}] ${summary}`);
    }
    lines.push("");
  }

  const typeCounts: Record<string, number> = {};
  for (const mem of projectMems) {
    const t = String((mem as Record<string, unknown>).memory_type ?? "OTHER").toUpperCase();
    typeCounts[t] = (typeCounts[t] ?? 0) + 1;
  }

  lines.push(`📊 Total: ${projectMems.length} entries across ${sortedMonths.length} month(s)`);
  const typeStr = Object.entries(typeCounts).sort((a, b) => b[1] - a[1]).map(([k, v]) => `${k}:${v}`).join(", ");
  lines.push(`   Types: ${typeStr}`);

  return lines.join("\n");
}

// ============================================================================
// Calendar View Formatter
// ============================================================================

function formatCalendarOutput(
  memories: MemoryNode[],
  semester: string,
  course: string | undefined,
  week: number | undefined,
  view: string
): string {
  const lines: string[] = [];

  if (course) {
    lines.push(`## 📅 ${semester} - ${course}`);
  } else {
    lines.push(`## 📅 ${semester} 学期笔记`);
  }
  lines.push("");

  if (memories.length === 0) {
    lines.push("_暂无笔记_");
    return lines.join("\n");
  }

  // Group by date
  const byDate: Record<string, MemoryNode[]> = {};
  for (const mem of memories) {
    const created = String((mem as Record<string, unknown>).created_at ?? (mem as Record<string, unknown>).timestamp ?? "");
    const dateStr = created ? created.slice(0, 10) : new Date().toISOString().slice(0, 10);
    if (!byDate[dateStr]) byDate[dateStr] = [];
    byDate[dateStr].push(mem);
  }

  const sortedDates = Object.keys(byDate).sort().reverse();

  if (view === "list") {
    lines.push("| 日期 | 类型 | 内容摘要 |");
    lines.push("|------|------|---------|");

    for (const dateStr of sortedDates.slice(0, 20)) {
      for (const mem of byDate[dateStr]) {
        const memType = String((mem as Record<string, unknown>).memory_type ?? (mem as Record<string, unknown>).type ?? "NOTE");
        let content = String((mem as Record<string, unknown>).content ?? mem.key ?? "").slice(0, 50);
        if (content.length === 50) content += "...";
        content = content.replace(/\|/g, "\\|").replace(/\n/g, " ");
        lines.push(`| ${dateStr} | ${memType} | ${content} |`);
      }
    }
  } else if (view === "week") {
    if (week) lines.push(`### 第 ${week} 周`);
    lines.push("");

    const weekdays = ["周一", "周二", "周三", "周四", "周五", "周六", "周日"];
    for (const dateStr of sortedDates.slice(0, 7)) {
      try {
        const dt = new Date(dateStr);
        const weekday = weekdays[(dt.getDay() + 6) % 7];
        lines.push(`**${dateStr} ${weekday}**`);
      } catch {
        lines.push(`**${dateStr}**`);
      }

      for (const mem of byDate[dateStr]) {
        const memType = String((mem as Record<string, unknown>).memory_type ?? (mem as Record<string, unknown>).type ?? "NOTE");
        const content = String((mem as Record<string, unknown>).content ?? mem.key ?? "").slice(0, 60).replace(/\n/g, " ");
        lines.push(`- [${memType}] ${content}`);
      }
      lines.push("");
    }
  } else {
    // month view
    lines.push("| 日期 | 笔记数 | 类型分布 |");
    lines.push("|------|--------|---------|");

    for (const dateStr of sortedDates.slice(0, 30)) {
      const mems = byDate[dateStr];
      const types: Record<string, number> = {};
      for (const m of mems) {
        const t = String((m as Record<string, unknown>).memory_type ?? "OTHER");
        types[t] = (types[t] ?? 0) + 1;
      }
      const typeStr = Object.entries(types).map(([k, v]) => `${k}:${v}`).join(", ");
      lines.push(`| ${dateStr} | ${mems.length} | ${typeStr} |`);
    }
  }

  const total = memories.length;
  const typeCounts: Record<string, number> = {};
  for (const mem of memories) {
    const t = String((mem as Record<string, unknown>).memory_type ?? "OTHER");
    typeCounts[t] = (typeCounts[t] ?? 0) + 1;
  }

  lines.push("");
  lines.push(`📊 总计: ${total} 条笔记`);
  if (Object.keys(typeCounts).length > 0) {
    const typeStr = Object.entries(typeCounts).sort((a, b) => b[1] - a[1]).map(([k, v]) => `${k}:${v}`).join(", ");
    lines.push(`   类型: ${typeStr}`);
  }

  return lines.join("\n");
}

// ============================================================================
// memos_calendar
// ============================================================================

export async function handleMemosCalendar(arguments_: Record<string, unknown>): Promise<TextContent[]> {
  const cubeId = getCubeIdFromArgs(arguments_);
  const mode = String(arguments_.mode ?? "student");
  const semester = String(arguments_.semester ?? "current");
  const course = arguments_.course ? String(arguments_.course) : undefined;
  const week = arguments_.week !== undefined ? Number(arguments_.week) : undefined;
  const view = String(arguments_.view ?? "list");

  const [regSuccess, regError] = await ensureCubeRegistered(cubeId);
  if (!regSuccess) return cubeRegistrationError(cubeId, regError);

  if (mode === "project") {
    const result = await apiCallWithRetry(
      "GET",
      `${MEMOS_URL}/memories`,
      cubeId,
      { params: { user_id: MEMOS_USER, mem_cube_id: cubeId, limit: 200 } },
      ensureCubeRegistered
    );

    if (result.success && result.data) {
      const resultData = (result.data as Record<string, unknown>).data as SearchData ?? {};
      const memories: MemoryNode[] = [];
      for (const cubeData of resultData.text_mem ?? []) {
        const memData = cubeData.memories;
        if (memData && !Array.isArray(memData) && memData.nodes) {
          memories.push(...memData.nodes);
        } else if (Array.isArray(memData)) {
          memories.push(...(memData as MemoryNode[]));
        }
      }
      const output = formatProjectTimeline(memories, cubeId);
      return [{ type: "text", text: output }];
    } else if (result.data) {
      return apiErrorResponse("Calendar", String((result.data as Record<string, unknown>).message ?? "Unknown error"));
    } else {
      return apiErrorResponse("Calendar", `HTTP ${result.status}`);
    }
  }

  // Student mode
  let [startDate, endDate] = getSemesterDates(semester);
  if (week !== undefined) {
    [startDate, endDate] = getWeekDates(startDate, week);
  }

  const result = await apiCallWithRetry(
    "GET",
    `${MEMOS_URL}/memories`,
    cubeId,
    { params: { user_id: MEMOS_USER, mem_cube_id: cubeId, limit: 100 } },
    ensureCubeRegistered
  );

  if (result.success && result.data) {
    const resultData = (result.data as Record<string, unknown>).data as SearchData ?? {};
    let memories: MemoryNode[] = [];
    for (const cubeData of resultData.text_mem ?? []) {
      const memData = cubeData.memories;
      if (memData && !Array.isArray(memData) && memData.nodes) {
        memories.push(...memData.nodes);
      } else if (Array.isArray(memData)) {
        memories.push(...(memData as MemoryNode[]));
      }
    }

    // Filter by course
    if (course) {
      const courseLower = course.toLowerCase();
      memories = memories.filter(
        (m) =>
          String((m as Record<string, unknown>).content ?? "").toLowerCase().includes(courseLower) ||
          String((m as Record<string, unknown>).tags ?? "").toLowerCase().includes(courseLower)
      );
    }

    // Filter by date range
    const filtered: MemoryNode[] = [];
    for (const mem of memories) {
      const created = String((mem as Record<string, unknown>).created_at ?? "");
      if (created) {
        try {
          const dt = new Date(created.replace("Z", "+00:00"));
          const dtNoTz = new Date(dt.getFullYear(), dt.getMonth(), dt.getDate());
          if (dtNoTz >= startDate && dtNoTz <= endDate) {
            filtered.push(mem);
          }
        } catch {
          filtered.push(mem);
        }
      } else {
        filtered.push(mem);
      }
    }

    const output = formatCalendarOutput(filtered, semester, course, week, view);
    return [{ type: "text", text: output }];
  } else if (result.data) {
    return apiErrorResponse("Calendar", String((result.data as Record<string, unknown>).message ?? "Unknown error"));
  } else {
    return apiErrorResponse("Calendar", `HTTP ${result.status}`);
  }
}
